docs/admin_settings.md
# Survey App Settings

There are a number of suvey app settings that the primary and survey admins can set via the 
Settings tab on the [Admin Dashboard](admin_dashboard.md).

These apply to the function of the app regardless of the current active survey.

These settings are:

**Look and Feel**

- **app_name** is the name of the survey app as it will appear
  to participants. It will be displayed in the navigation
  bar when there is no active survey and will appear as the
  from line in any email sent out by the survey.  If left
  unset, this will default to `Time and Talent Survey`.
- **app_logo** allows you to select an image to be displayed
  in the navigation bar before the app/survey name.  This
  can be most (any?) standard image format. It is recommended,
  however, that it be roughly square in shape with a resolution
  of at least 128x128. If left unset, no image will be shown.
- **timezone** sets the timezone used to display any time in
  the survey, the dashboard, or emails.  Internally, all
  times are recorded in UTC so that hanging the timezone 
  does not impact how times are written to the database.
  If left unset, the timezone will default to that used
  by your HTTP server.

**Admin**

- **admin_name** is the real (or pseudo) name for the site 
  admin defined in the config .ini file.  This will only
  be used if no participant (user) has been assigned the
  role of `primary admin`.  If left unset, the site admin
  name will be shown as `the survey admin`.
- **admin_email** is the email address displayed when instructing
  participants to contact an admin due to a problem with the
  app.  This will only be used if no participant has been
  assigned the role of `primary admin`.  If left unset, 
  no email address will displayed.

**Logging**

- **log_file** is the full path to the location of the 
  app's log file.  This displays the value set in the
  config .ini file.  It cannot be changed in the dashboard.
- **log_level** sets the level of information to include in
  the logging (errors, warnings, info, dev). The suggested
  level is errors and warnings unless you are doing active
  app development.

**Password Reset**

- **pwreset_timeout** is the number of minutes that the token
  issued during a password reset request will be valid before
  it times out.  If left unset, this defaults to 15 minutes.
- **pwreset_length** is the number of characters that will be
  included in the generated password reset tokens. If left
  unset, this defaults to 10 characters.

*Note that password reset settings only apply for particpants
that have provided an email address and can thus request a 
password reset.*

**Email Reminders**

- **reminder_freq** is the minimum time (hours) between 
  any reminder email sent to any given participant. This
  is set to guard against inadvertant spamming if any
  admin initiates the sending of reminders to one or
  more set of participants.

**Email Server**

*I will not be able to help you determine your particular
settings.  You will need to work with your email service
provider to work out the values for their SMTP servers.
I am using gmail and there is a lot of good documentation
out there on setting this up.*

- **smtp_host** is the URL for the SMTP server
- **smtp_auth** SMTPS or STARTTLS
- **smtp_port** will probably be either 465 for SMTPS or 587 for STARTTLS
- **smtp_username** is a SMTP login credential
- **smtp_password** is a SMTP login credential
- **smtp_reply_email** is an optional SMTP used to set the reply-to 
  field in any outgoing email.
- **smtp_reply_name** is an optional SMTP used to set the reply-to 
  field in any outgoing email.
- **smtp_debug** is the only setting not associated with your SMTP
  server. It tells the survey app how to log any debug responses
  it receives from the SMTP server. If any are selected for logging,
  they will be logged and the `info` (aka `notices`) level. 

After setting email server settings, you can then send a test email 
to the primary (or site) admin to verify you have everything set correctly.

include/settings.php
<?php
namespace tlc\tts;

if(!defined('APP_DIR')) { http_response_code(405); error_log("Invalid entry attempt: ".__FILE__); die(); }

require_once(app_file('include/db.php'));

class Settings {

  static private $defaults = [
    'app_name'        => 'Time and Talent Survey',
    'timezone'        => 'UTC',
    'is_dev'          => false,
    'admin_name'      => 'the survey admin',
    'pwreset_timeout' => 15,
    'pwreset_length'  => 10,
    'reminder_freq'   => 24, // hours
    'log_level'       => 2,
    'smtp_auth'       => 1,  // 0=SMTPS, 1=STARTTLS
    'smtp_debug'      => 0,  // 0=None, 1=Server->Client, 2=Server<->Client, 3=extra
    'summary_flags'   => 0,
  ];

  static private $values = array();

  static public function load_all()
  {
    $rows = MySQLSelectValues('select name,value from tlc_tts_settings');

    foreach($rows as $row) {
      $values[$row[0]] = $row[1];
    }
  }

  static public function update(...$kv) {
    if(count($kv) == 1) {
      $kv = $kv[0];
      foreach($kv as $k=>$v) {
        self::set($k,$v);
      }
    } else {
      while(count($kv)>1) {
        $k = array_shift($kv);
        $v = array_shift($kv);
        self::set($k,$v);
      }
    }
  }

  static public function validate(...$args) { 
    if( count($args) == 1 ) {
      $kv = $args[0];
    } else {
      $kv = array();
      while(count($args) > 1) {
        $k = array_shift($args);
        $v = array_shift($args);
        $kv[$k] = $v;
      }
    }
    $errors = array();
    $error='';
    foreach($kv as $key=>$value) {
      $vfunc = "tlc\\tts\\validate_$key";
      if(function_exists($vfunc)) {
        if(!$vfunc($value,$error)) { $errors[$key] = $error; }
      }
    }
    return $errors;
  }

  static public function default($key) 
  { 
    return self::$defaults[$key] ?? null; 
  }
  
  static public function raw($key) 
  {
    if(key_exists($key,self::$values)) { return self::$values[$key]; }

    $value = MySQLSelectValue('select value from tlc_tts_settings where name=?','s',$key); 
    if( $value !== null && $value !== '' ) {
      self::$values[$key] = $value;
    }
    return $value;
  }

  static public function get($key) 
  {
    if(key_exists($key,self::$values)) { return self::$values[$key]; }

    $value = MySQLSelectValue('select value from tlc_tts_settings where name=?','s',$key); 
    if( $value !== null && $value !== '' ) {
      self::$values[$key] = $value;
    } else {
      $value = self::$defaults[$key] ?? null; 
    }
    return $value;
  }

  static public function clear($key)
  {
    self::set($key,null);
  }

  static public function set($key,$value) 
  {
    if(is_null($value) || $value==='') {
      unset(self::$values[$key]);
      MySQLExecute('delete from tlc_tts_settings where name=?','s',$key);
    }
    else {
      MySQLExecute(
        "insert into tlc_tts_settings (name,value) values (?,?) on duplicate key update value=?",
        'sss',$key,$value,$value
      );
    }
  }
};

Settings::load_all();

// fix timezone
date_default_timezone_set(Settings::get('timezone'));
MySQLExecute("SET time_zone = '".date('P')."'");

//
// Convenience Accessors
//

function get_setting($key)        { return Settings::get($key);      } 
function set_setting($key,$value) { Settings::set($key,$value);      }
function clear_setting($key)      { Settings::clear($key);           } 
function setting_default($key)    { return Settings::default($key);  }

// App Look-and-Feel settings
function app_name()            { return get_setting('app_name'); }
function timezone()            { return get_setting('timezone'); }
function is_dev()              { return get_setting('is_dev'); }

// Admin settings
function admin_name()          { return get_setting('admin_name'); }
function admin_email()         { return get_setting('admin_email'); }
function primary_admin()       { return get_setting('primary_admin'); }

function admin_contact() {
  $contact = htmlspecialchars(admin_name(), ENT_QUOTES, 'UTF-8');
  $email = admin_email();
  if ($email) {
      $email_escaped = htmlspecialchars($email, ENT_QUOTES, 'UTF-8');
      $contact = sprintf("<a href='mailto:%s'>%s</a>", $email_escaped, $contact);
  }
  return $contact;
}

// Password reset settings
function pwreset_timeout()  { return get_setting('pwreset_timeout'); } // minutes
function pwreset_length()   { return min(20,max(4, get_setting('pwreset_length'))); }
function reminder_freq()    { return get_setting('reminder_freq'); } // hours

// Logging settings
function log_level()        { return get_setting('log_level', 2);  }

// SMTP settings
function smtp_host()        { return get_setting('smtp_host'); }
function smtp_auth()        { return get_setting('smtp_auth'); }
function smtp_username()    { return get_setting('smtp_username'); }
function smtp_password()    { return get_setting('smtp_password'); }
function smtp_reply_email() { return get_setting('smtp_reply_email'); }
function smtp_reply_name()  { return get_setting('smtp_reply_name'); }
function smtp_debug()       { return get_setting('smtp_debug'); }

function smtp_port() { 
  $port = get_setting('smtp_port');
  if(!$port) {
    // if port is not set, infer it from smtp_auth
    //   STARTTLS => 587
    //   SMTPS    => 465
    $port = smtp_auth() ? 587 : 465;
  }
  return $port;
}

//
// Date and Time
//

date_default_timezone_set(timezone());

//
// Validation functions
//

function _fix_validate_value(&$value)
{
  if(isset($value)) { $value = trim($value); }
  else              { $value = ''; }
}

function validate_timezone($timezone,&$error=null) {
  $error = '';
  _fix_validate_value($timezone);
  if($timezone==='') { return true; }
  if(!date_default_timezone_set($timezone)) { 
    $error = "unrecognized timezone";
  }
  return strlen($error) == 0;
}

function validate_admin_name($name, &$error = null): bool {
  $error = '';
  _fix_validate_value($name);
  if($name==='') { return true; }
  if (!preg_match("/^[\p{L}\p{N} .'_-]+$/u", $name)) {
      $error = "Admin name contains invalid characters";
  }
  return strlen($error) == 0;
}

function validate_admin_email($email,&$error=null) {
  $error = '';
  _fix_validate_value($email);
  if($email==='') { return true; }
  if(!filter_var($email, FILTER_VALIDATE_EMAIL)) {
    $error = "invalid email address"; 
  }
  return strlen($error) == 0;
}

function validate_pwreset_timeout($timeout,&$error=null) {
  $error = '';
  _fix_validate_value($timeout);
  if($timeout==='') { return true; }
  if(is_numeric($timeout)) {
    $timeout = 1*$timeout;
    if($timeout <= 0) { $error = "not a postive number"; }
  } else { 
    $error = "not a number";
  }
  return strlen($error) == 0;
}

function validate_pwreset_length($len,&$error=null) {
  $error = '';
  _fix_validate_value($len);
  if($len==='') { return true; }
  if(is_numeric($len)) {
    $len = 1*$len;
    if(!is_integer($len)) { $error = "not an integer"; }
    elseif($len<4)        { $error = "too small (<4)"; }
    elseif($len>20)       { $error = "too big (>20)"; }
  } else {
    $error = "not an integer";
  }
  return strlen($error) == 0;
}

function validate_smtp_host($host,&$error=null) {
  $error = '';
  _fix_validate_value($host);
  if( $host === '' ) { $error = "missing"; }
  elseif(!filter_var($host,FILTER_VALIDATE_DOMAIN)) {
    $error = "invalid domain name";
  }
  return strlen($error) == 0;
}

function validate_smtp_username($name,&$error=null) {
  $error = '';
  _fix_validate_value($name);
  if($name==='') { $error = 'missing'; }
  return strlen($error) == 0;
}

function validate_smtp_password($password,&$error=null) {
  $error = '';
  _fix_validate_value($password);
  if($password==='') { $error = 'missing'; }
  return strlen($error) == 0;
}

function validate_smtp_port($port,&$error=null) {
  $error = '';
  _fix_validate_value($port);
  if($port==='') { return true; }
  if(is_numeric($port)) {
    $port = 1*$port;
    if(!is_integer($port)) { $error = "not an integer"; }
    elseif($port<=0)       { $error = "not a positive integer"; }
  } else {
    $error = "not an integer";
  }
  return strlen($error) == 0;
}

function validate_smtp_reply_email($email,&$error=null) {
  $error = '';
  _fix_validate_value($email);
  if($email==='') { return true; }
  if(!filter_var($email, FILTER_VALIDATE_EMAIL)) {
    $error = "invalid email address"; 
  }
  return strlen($error) == 0;
}
